tests/test.py

```diff
@@ -0,0 +1,185 @@
+
+log_file = open("test.log", 'w')
+
+
+def log(s):
+    log_file.write("%s\n" % s)
+
+
+class MavlinkSender(Thread):
+    '''MAVLink message sender'''
+
+    def __init__(self, name, port, sysId, compId, targetSysId, targetCompId):
+        super().__init__()
+        self.name = name
+        self.sysId = sysId
+        self.compId = compId
+        self.targetSysId = targetSysId
+        self.targetCompId = targetCompId
+        self.mav = mavutil.mavlink_connection(
+            'udpout:127.0.0.1:' + str(port), source_system=self.sysId)
+        self.sender_thread = Thread(target=self.send_loop)
+        self.received = []
+        self.success = True
+
+    def send_loop(self):
+        for i in range(0, 10):
+            self.mav.mav.ping_send(
+                int(time.time() * 1000), i, self.targetSysId,
+                self.targetCompId)
+            sleep(0.5)
+
+    def run(self):
+        self.sender_thread.start()
+
+        while (True):
+            msg = self.mav.recv_match(blocking=True, timeout=5)
+            if msg is not None:
+                if self.targetSysId != 0 and msg.get_srcSystem(
+                ) != self.targetSysId:
+                    log("Received unexpected response from %d/%d - current target: %d/%d"
+                        % (msg.get_srcSystem(), msg.get_srcComponent(),
+                           self.targetSysId, self.targetCompId))
+                    self.success = False
+
+                # TODO maybe check if no message sent has seq bigger than any sent?
+                self.received.append(msg)
+            else:
+                break
+
+    def accept(self, func):
+        return self.success and func(self.received)
+
+
+class MavlinkReceiver(Thread):
+    '''MAVLink message receiver'''
+
+    def __init__(self, name, port, sysId, compId):
+        super().__init__()
+        self.name = name
+        self.sysId = sysId
+        self.compId = compId
+        self.mav = mavutil.mavlink_connection(
+            'udpin:127.0.0.1:' + str(port), source_system=self.sysId)
+        self.received = []
+        self.success = True
+
+    def run(self):
+        while (True):
+            msg = self.mav.recv_match(blocking=True, timeout=5)
+            if msg is not None:
+                self.received.append(msg)
+                if msg.target_system not in [0, self.sysId]:
+                    log("Received unexpected message from: %d/%d, meant to %d/%d - current receiver: %d/%d"
+                        % (msg.get_srcSystem(), msg.get_srcComponent(),
+                           msg.target_system, msg.target_component, self.sysId,
+                           self.compId))
+                    self.success = False
+                self.mav.mav.ping_send(
+                    int(time.time() * 1000), msg.seq,
+                    msg.get_srcSystem(), msg.get_srcComponent())
+            else:
+                break
+
+    def accept(self, func):
+        return self.success and func(self.received)
+
+
```

Comment:
It is a good habit to add:
if __name__ == "__main__":
to check if the script is being run or if it is being included. If the idea is to run the code even if it is being included, ignore that. hehe